Source/Core/Core/HW/GCMemcard/GCMemcardRaw.h

```diff
@@ -30,7 +30,7 @@ class MemoryCard : public MemoryCardBase
   void DoState(PointerWrap& p) override;
 
 private:
-  bool IsAddressInBounds(u32 address) const { return address <= (m_memory_card_size - 1); }
+  bool IsAddressInBounds(u32 address, u32 length) const { return address + length <= (m_memory_card_size - 1); }
```

Comment:
What if `address + length` overflows?

Reply:
With the length included we want to get rid of -1 as well. If hypothetically the card had 10 valid addresses and we wanted to read all of them then address == 0, length == 10, m_memory_card_size == 10, and the check 10 <= 9 fails.

Reply:
Yeah that's a valid concern. 
What do you think about changing the check like that?
```cpp
  bool IsAddressInBounds(u32 address, u32 length) const 
  {
    u64 end_address = address;
    end_address += length;
    return end_address <= (m_memory_card_size - 1); 
  }
  ```

Reply:
Also do you know why the length can be negative?

Reply:
As far as I can tell there's no reason for length to be negative; it just was done that way in when the abstraction for memory card folders were created in 404e9ce3ee722872b941eefb24d87e95dc21b55b. `Read` and `Write` are both only called from `EXI_DeviceMemoryCard.cpp` either with a `u32` parameter (from `DMARead` or `DMAWrite`), the constant `1`, or `static_cast<s32>(header.size())` (where `header` is `std::array<u8, 20>` so that's just the constant `20`). It doesn't make sense either.

(`GCMemcardDirectory`'s `m_last_block` is `s32` because that uses `-1` as a sentinel value but I don't think that affects things here. The other `s32` usage there is a bit dubious still.)

I think the check might be a bit more obvious like this, but there's a ton of ways to nitpick about it probably:

```c++
  bool IsAddressInBounds(u32 address, u32 length) const 
  {
    u64 end_address = static_cast<u64>(address) + static_cast<u64>(length);
    return end_address <= static_cast<u64>(m_memory_card_size);
  }
```